### PR TITLE
Audit Command for Resource Compliance

### LINKED ISSUE #372: Auditing Compliance / Completion of Data Map Attributes
### Is your feature request related to a specific problem?

With so many attributes and potential need for manual attribution a method to score/measure compliance or completeness of the data map would be a useful tool for fides users.

### Describe the solution you'd like

A CLI command to audit completeness of the combined data map model (`scan` command potentially?) and provide not only an overview of the completeness but also actionable items as required to have full coverage.

### Describe alternatives you've considered, if any

n/a

### Additional context

This has been discussed internally at a number of points and seems to fit well to ship with the run state of the data map. Having something like this will be a huge boost in providing clarity around what is required as well as (potentially) who has the action to complete.

### PR DESCRIPTION
**Pull request by SteveDMurphy**
Closes #372

### Code Changes

* [x] Add a new option to `--include-audit` of server resources when running `evaluate`
* [x] Add validation of a systems attributes that support the data map
* [x] Add validation for Organization attributes
* [x] ~Add validation of a dataset attribute~ not required at this time
* [ ] Remove the explicit `audit` command left from earlier

### Steps to Confirm

* [x] `fidesctl evaluate demo_resources/ -a` successfully audits and returns valid and clear(er) messaging

### Pre-Merge Checklist

* [x] All CI Pipelines Succeeded
* Documentation Updated:
  - [x] documentation complete, or draft/outline provided (tag docs-team to complete/review on this branch)
  - [ ] documentation issue created (tag docs-team to complete issue separately)
* [x] Issue Requirements are Met
* [x] Relevant Follow-Up Issues Created
* [x] Update `CHANGELOG.md`

### Description Of Changes

A new option to audit for compliance seemed to make sense here as a starting point. Long-term, the vision is to extend policies to handle multiple types of validation (ci checks, governance, etc.). For this reason, combining with `evaluate` instead of creating another command.

Steps to validate:
1. start up the cli with `nox -s cli`
1. `fidesctl db reset -y`
1. `fidesctl evaluate demo_resources/ -a`
1. (should return an error when auditing systems)
1. set both data_uses in `demo_system.yml` to `third_party_sharing.personalized_advertising.direct_marketing`
1. set both data_subjects in `demo_system.yml` to `potential_customer`
1. `fidesctl evaluate demo_resources/ -a` (should now succeed)

### CODE REVIEW

**Comment by SteveDMurphy** [comment]
@ThomasLaPiana could I get an early opinion on if this is what we were thinking for the initial auditing of resources? As part of adding more validation for a system (currently just 1 item) I was thinking about having this return a dictionary of findings

Thanks!

File: src/fidesctl/cli/commands/audit.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+
+@audit.command(name="systems")
+@click.pass_context
+@with_analytics
+def audit_system(
+    ctx: click.Context,
+) -> None:
+    """
+    Export a system in a data map format.
```
Comment (ThomasLaPiana) [inline src/fidesctl/cli/commands/audit.py:23]: copy pasta 🙈

**Comment by ThomasLaPiana** [comment]
@SteveDMurphy I love the direction this is going! I'll add a few nits but overall I think its ready to ship. Is there anything you think its missing?

File: src/fidesctl/cli/commands/audit.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
+# """Contains the export group of CLI commands for Fidesctl."""
```
Comment (ThomasLaPiana) [inline src/fidesctl/cli/commands/audit.py:1]:
```suggestion
# """Contains the audit group of CLI commands for Fidesctl."""
```

File: src/fidesctl/core/audit.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        url, headers, "system", exclude_keys=exclude_keys
+    )
+    audit_findings = 0
+    for system in system_resources:
+        print(f"Auditing System: {system.name}")
+        new_findings = validate_system_attributes(system, url, headers)
+        audit_findings = audit_findings + new_findings
+
+    if audit_findings > 0:
+        print(f"{audit_findings} issues were detected in auditing system completeness.")
```
Comment (ThomasLaPiana) [inline src/fidesctl/core/audit.py:35]:
```suggestion
        print(f"{audit_findings} issue(s) were detected in auditing system completeness.")
```

**Comment by SteveDMurphy** [comment]
Thanks for the in progress check and updates @ThomasLaPiana ! Glad to hear 🙌🏽 

I think just some refactoring for the checks/output, and potentially getting ahead of how we want to return audit findings. In the end this is definitely just a first step towards some semblance of auditing so also not feeling the need to overdo it too much 😄

**Comment by SteveDMurphy** [comment]
@conceptualshark I am not sure if we wanted to add something else docs-wise for this or maybe a separate issue? tried to beef up the descriptions that will populate with the CLI docs but adding you as a reviewer if that's ok to give an opinion on whether or not we want to do anything else 👍🏽

**Comment by conceptualshark** [comment]
@SteveDMurphy I'd like to call this out individually in the data map page if it will help people identify any missing values they need to add in. Is this the case? If it's only `administrating_department` right now and more are going to come, maybe we could finalize the ticket then.

**Comment by SteveDMurphy** [comment]
> @SteveDMurphy I'd like to call this out individually in the data map page if it will help people identify any missing values they need to add in. Is this the case? If it's only `administrating_department` right now and more are going to come, maybe we could finalize the ticket then.
> 
> Either way - it can be its own issue separate to this!

Thanks @conceptualshark !! Added a few more attribute checks in here so added a section to the data map guide. Should be tagged for a review now as well, thanks again for the quick response/thoughts so far 🙌🏽

**Comment by ThomasLaPiana** [comment]
@SteveDMurphy is this ready for another review?

**Comment by SteveDMurphy** [comment]
> @SteveDMurphy is this ready for another review?

Yep! Sorry I forgot to hit the re-request review from your earlier comment

**Comment by ThomasLaPiana** [comment]
@SteveDMurphy did the `steps to validate` change? I'm getting very different output:

```root@0c4f9ad29f51:/fides# fidesctl apply demo_resources/
Loading resource manifests from: demo_resources/
Taxonomy successfully created.
----------
Processing data_use resource(s)...
APPLIED 1 data_use resource(s).   
----------
Processing dataset resource(s)... 
APPLIED 1 dataset resource(s).  
----------
Processing system resource(s)...
APPLIED 2 system resource(s).   
----------
Processing policy resource(s)...
APPLIED 1 policy resource(s).
----------
Processing organization resource(s)...
APPLIED 1 organization resource(s).
----------
Processing registry resource(s)... 
APPLIED 1 registry resource(s).       
----------
Processing data_subject resource(s)...
APPLIED 1 data_subject resource(s).   
----------
root@0c4f9ad29f51:/fides# fidesctl audit systems
Auditing System: Demo Analytics System
improve.system missing recipients in Demo Analytics System.
improve.system missing legal_basis in Demo Analytics System.
improve.system missing special_category in Demo Analytics System.
customer missing rights in Demo Analytics System.
customer missing automated_decisions_or_profiling in Demo Analytics System.
Auditing System: Demo Marketing System
advertising missing recipients in Demo Marketing System.      
advertising missing legal_basis in Demo Marketing System.     
advertising missing special_category in Demo Marketing System.
customer missing rights in Demo Marketing System.
customer missing automated_decisions_or_profiling in Demo Marketing System.
10 issue(s) were detected in auditing system completeness.
```

**Comment by SteveDMurphy** [comment]
> @SteveDMurphy did the `steps to validate` change?

@ThomasLaPiana Yes, sorry! Let me update that quickly in the description. Adding the rest of the attributes in here make it go a bit deeper. Essentially, adding the extended taxonomy in there for `data_use` and `data_subject` (I believe, confirming now)

**Comment by SteveDMurphy** [comment]
> Essentially, adding the extended taxonomy in there for `data_use` and `data_subject` (I believe, confirming now)

Confirmed that was the issue, added those steps in the PR description. My bad again on that!

It also look like the `administrating_department` change got committed at some point, we could probably remove that as well but not sure how much that matters either way.

**Comment by ThomasLaPiana** [comment]
@SteveDMurphy got it working, thank you!

**Review by ThomasLaPiana — DISMISSED** [review]
LGTM! Tested locally

**Comment by SteveDMurphy** [comment]
thanks @ThomasLaPiana !! 

After some conversations around naming this week with different folks (`generate` vs `scan` are one example) I am having second thoughts around the `audit` term being mixed in...

The term `audit` could easily be used in place of `scan`, `scan` is already being confused with `generate`, and `generate` is functionally correct for what we are doing with the data we retrieve at the end but still part of this current group of items that could mean different things.

Any thoughts here on naming or is this conversation better suited elsewhere maybe? 🙏🏽

**Comment by ThomasLaPiana** [comment]
@SteveDMurphy i totally see your point here....

why not make it a flag? We've talked about having a "strict" flag in the past, and maybe this is the perfect place to add it in...

Why make the user run yet another command instead of just `fidesctl evaluate --strict` or something?

**Comment by SteveDMurphy** [comment]
Thanks @ThomasLaPiana , yea I can see a flag being another option here 🤔  

One concern with `evaluate`  could be in diluting the main purpose there, which is to ensure compliance against a policy. I'd almost rather surface this as a warning during the `apply` (potentially?) so the user can see when applying something that certain attributes are still needed or even the `export` command itself. 

Alternatively, maybe still having a separate command but rethinking some of our current naming to have clearer separation of duties?

**Comment by ThomasLaPiana** [comment]
> 

Let's try to give an imagined real world example and compare use-cases:

1. Keep it as part of the export command: If I were part of the compliance team, I would be confused and frustrated if I tried to export a datamap only to learn that I hadn't actually configured it properly, or worse yet it was configured properly but an engineer messed up the configuration. I've then got to go talk to the eng team to get it fixed, merged, and then try again. Due to this incident, we now add the `fidesctl export datamap` command to the CI check so this doesn't happen again
2. Add it as part of `evaluate` via a flag: This keeps the engineer only needing to run that one command in CI (as opposed to running both `fidesctl evaluate` and `fidesctl export datamap`)
3. Adding the flag to `apply`: This is functionally the exact same as adding it to `evaluate` since `apply` runs as part of `evaluate`.

With 2 and 3, maybe we're overloading those commands by adding that flag? I'm not sure. It does seem weird that `fidesctl audit` doesn't audit the entire thing though, and requires separate commands. I'd think if you run `fidesctl audit` it'd return a list of _all_ errors, not just specific ones. This saves you having to run the command twice to make sure you're actually compliant for a data map.

Just some food for thought, lmk what your thoughts are :)

File: tests/core/test_dataset.py [hunk 4]
```diff
@@ -314,11 +315,13 @@ def test_find_missing_datasets() -> None:
 # External databases require credentials passed through environment variables
 SNOWFLAKE_URL_TEMPLATE = "snowflake://FIDESCTL:{}@ZOA73785/FIDESCTL_TEST"
 SNOWFLAKE_URL = SNOWFLAKE_URL_TEMPLATE.format(
-    os.getenv("SNOWFLAKE_FIDESCTL_PASSWORD", "")
+    quote_plus(os.getenv("SNOWFLAKE_FIDESCTL_PASSWORD", ""))
```
Comment (SteveDMurphy) [inline tests/core/test_dataset.py:318]: @ThomasLaPiana this resolves the failing external pytest error, will search through issues now and either add a note or create a new one

**Comment by ThomasLaPiana** [comment]
@SteveDMurphy sounds good. Now that the fix is with this PR though, it forces us to make a decision quickly and ship this PR to get the fix. Can we open a quick smaller bug fix PR so we can get more feedback on this from the team at tomorrow's planning?

**Comment by SteveDMurphy** [comment]
> Now that the fix is with this PR though, it forces us to make a decision quickly and ship this PR to get the fix

I have a bad habit of doing this 🤦🏽 fix moved over to #661 - thanks @ThomasLaPiana 👍🏽

**Comment by SteveDMurphy** [comment]
I feel like this is at least closer to what was discussed in office hours recently. The only hangup for me is if it does become a flag for `evaluate` we should change this to also only look at resources in the specified manifest directory (while the `audit` currently looks at all resources on the server). There is a function for both in my mind so a bit torn

**Comment by SteveDMurphy** [comment]
@ThomasLaPiana I am just going to review the docs once more to clean up any of the original changes referenced in there but functionally I think this is good for another look 👍🏽  

I tried to step the commits through in some sort of a sequence so we can see what was changed etc. to make this happen. Let me know what you think, thanks again for the help/patience!

File: src/fidesctl/cli/commands/core.py [hunk 5]
```diff
@@ -48,13 +49,20 @@ def apply(ctx: click.Context, dry: bool, diff: bool, manifests_dir: str) -> None
     "--message",
     help="A message that you can supply to describe the context of this evaluation.",
 )
+@click.option(
+    "-a",
+    "--include-audits",
```
Comment (ThomasLaPiana) [inline src/fidesctl/cli/commands/core.py:54]:
```suggestion
    "--audit",
```

Just a nit, I think `include-audit` sounded a little weird and verbose without being descriptive.

File: src/fidesctl/cli/commands/core.py [hunk 6]
```diff
@@ -48,13 +49,20 @@ def apply(ctx: click.Context, dry: bool, diff: bool, manifests_dir: str) -> None
...
+    is_flag=True,
+    help="Include optional audits for data map attribute compliance.",
```
Comment (ThomasLaPiana) [inline src/fidesctl/cli/commands/core.py:56]:
```suggestion
    help="Raise errors if resources are missing attributes required for building a data map.",
```

**Review by ThomasLaPiana — APPROVED** [review]
LGTM! Made a few small nits

Thanks for getting feedback on this and working together to get this shipped

File: src/fidesctl/cli/commands/core.py (hunk 5 above)
Comment (SteveDMurphy) [reply src/fidesctl/cli/commands/core.py:54]: ya I had a tough time with naming, brevity over descriptive etc. It's sort of an odd fit but happy to go with this for sure 👍🏽

**Pull request merged**